graph_worker.cpp

```diff
@@ -19,37 +29,74 @@ void GraphWorker::startWorkers(Graph *_graph, BufferTree *_bf) {
 	}
 }
 
-void GraphWorker::stopWorkers() {
+void GraphWorker::stop_workers() {
 	shutdown = true;
+	workers[0]->bf->bypass_wait(true); // make the GraphWorkers bypass waiting in queue
+	pause_condition.notify_all();      // tell any paused threads to continue and exit
 	for (int i = 0; i < num_groups; i++) {
 		delete workers[i];
 	}
 	delete workers;
 }
 
+void GraphWorker::pause_workers() {
+	paused = true;
+	workers[0]->bf->bypass_wait(true); // make the GraphWorkers bypass waiting in queue
+
+	// wait until all GraphWorkers are paused
+	std::unique_lock<std::mutex> lk(pause_lock);
+	pause_condition.wait(lk, []{
+		for (int i = 0; i < num_groups; i++)
+			if (!workers[i]->get_thr_paused()) return false;
+		return true;
+	});
+	lk.unlock();
+}
+
+void GraphWorker::unpause_workers() {
+	workers[0]->bf->bypass_wait(false); // buffer-tree operations should block when necessary
+	paused = false;
+	pause_condition.notify_all();       // tell all paused workers to get back to work
+}
+
+/***********************************************
+ ************** GraphWorker class **************
+ ***********************************************/
 GraphWorker::GraphWorker(int _id, Graph *_graph, BufferTree *_bf) :
-  id(_id), graph(_graph), bf(_bf), thr(startWorker, this) {
+  id(_id), graph(_graph), bf(_bf), thr(start_worker, this) {
 }
 
 GraphWorker::~GraphWorker() {
-	if(id == 0)
-		bf->bypass_wait(); // to avoid race condition on shutdown. Tell threads to recheck wait
-	
-	// when a graph worker is deleted we must wait for the associated
-	// thread to finish its work. See stopWorkers and doWork
+	// join the GraphWorker thread to reclaim resources
 	thr.join();
 }
 
-void GraphWorker::doWork() {
+void GraphWorker::do_work() {
 	data_ret_t data;
 	while(true) {
-		// call get_data which will handle waiting on the queue
-		// and will enforce locking.
-		bool valid = bf->get_data(data);
-
-		if (valid)
-			graph->batch_update(data.first, data.second);
-		else if(shutdown)
+		if(shutdown)
 			return;
+		pause_lock.lock();
+		thr_paused = true; // this thread is currently paused
```

Comment:
I'm concerned this ordering of operations might cause us some problems when it comes to querying. if we push a set amount of data into the buffer tree and flush it, won't some tasks get stuck in the queue when we pause? for instance if the circular queue isn't empty.

Reply:
So this outer loop only runs when the thread pauses. The thread only pauses after failing to get data from the circular queue and if paused==true.

This means that so long as pause_workers is only called after force_flush that we should process all of the updates in the tree.

Reply:
ok, and this is ok because we can be reasonably confident that force_flush will return after the inner loop has already begun?

Reply:
The inner loop should begin before force_flush is even called and if there's anything in the buffer tree when it is the inner loop must be executing in order to process that data.

Reply:
but in the case that we only have a few updates to process, we’ll rely on the force_flush timing right?

Reply:
No because get_data will be a blocking call until we pause the GraphWorkers

Reply:
I’m a little confused. What’s the flow gonna be when, for instance, we push 10 updates into the buffer tree and then call `connected_components()`?

Reply:
Actually never mind i think i see it:
- we’ll call `force_flush`, and that puts some things into the circular queue
- the graph workers are notified and start pulling things out of the queue
- `force_flush` returns to the main thread
- we call pause, which blocks until all the graph workers are done

Reply:
Yes exactly